Rebus.AmazonS3/AmazonS3/AmazonS3SubscriptionsStorage.cs

```diff
@@ -0,0 +1,230 @@
+
+            HashSet<string> subscribers;
+            if(subscriptions.TryGetValue(topic, out subscribers))
+                return subscribers.ToArray();
+
+            return new string[0];
+        }
+
+        public async Task RegisterSubscriber(string topic, string subscriberAddress)
+        {
+            await EnsureBucketExist();
+
+            var subscriptions = await GetSubscriptions(topic);
+
+            subscriptions
+                .GetOrAdd(topic, () => new HashSet<string>())
+                .Add(subscriberAddress);
+
+            await SaveToS3(subscriptions, topic);
+        }
+
+        public async Task UnregisterSubscriber(string topic, string subscriberAddress)
+        {
+            await EnsureBucketExist();
+
+            var subscriptions = await GetSubscriptions(topic);
+
+            subscriptions
+                .GetOrAdd(topic, () => new HashSet<string>())
+                .Remove(subscriberAddress);
+
+            await SaveToS3(subscriptions, topic);
+        }
+
+        private async Task<Dictionary<string, HashSet<string>>> GetSubscriptions(string topic)
+        {
+            if (await ObjectExists(topic))
+            {
+                string jsonText = await FetchFromS3(topic);
+
+                var subscriptions = JsonConvert.DeserializeObject<Dictionary<string, HashSet<string>>>(jsonText);
+
+                return subscriptions;
+            }
+
+            return new Dictionary<string, HashSet<string>>();
+        }
+
+        private async Task<bool> ObjectExists(string key)
+        {
+            using (IAmazonS3 s3Client = CreateS3Client())
+            {
+                try
+                {
+                    await s3Client.GetObjectMetadataAsync(_options.BucketName, key);
+                    return true;
+                }
+                catch (AmazonS3Exception e)
+                {
+                    if (e.StatusCode == HttpStatusCode.NotFound)
+                    {
+                        return false;
+                    }
+
+                    throw new RebusApplicationException(e, "Unexpected exception occured");
+                }
+                catch (Exception e)
+                {
+                    throw new RebusApplicationException(e, "Unexpected exception occured");
+                }
+            }
+        }
+
+        private async Task<string> FetchFromS3(string key)
+        {
+            var request = new GetObjectRequest
+            {
+                BucketName = _options.BucketName,
+                Key = key
+            };
+
+            using (IAmazonS3 s3Client = CreateS3Client())
+            {
+                try
+                {
+                    using (var response = await s3Client.GetObjectAsync(request))
+                    {
+                        using (StreamReader reader = new StreamReader(response.ResponseStream))
+                        {
+                            return reader.ReadToEnd();
```

Comment:
Could probably be replaced by  `return await reader.ReadToEndAsync()` as not to block the calling thread while fetching the object text